uint8 mainPhase
uint8 subPhase